enterprise/backend/src/metabase_enterprise/enhancements/ee_strategy_impl.clj

```diff
@@ -5,14 +5,13 @@
   In OO pattern terminology, this is an implementation of the strategy pattern -- the implementation of the interface
   is determined at runtime."
   (:require [clojure.string :as str]
-            [metabase.public-settings.metastore :as settings.metastore]
             [pretty.core :refer [PrettyPrintable]]))
 
 (defn invoke-ee-when-enabled
-  "Impo for `reify-ee-strategy-impl`. Invoke `method` using `ee-impl` if EE features are enabled, otherwise invoke with
+  "Impl for `reify-ee-strategy-impl`. Invoke `method` using `ee-impl` if EE features are enabled, otherwise invoke with
   `oss-impl`."
-  [method ee-impl oss-impl & args]
-  (let [impl (if (settings.metastore/enable-enhancements?)
+  [enable-pred-var method ee-impl oss-impl & args]
+  (let [impl (if (enable-pred-var)
```

Comment:
this machinery is really clever. I think the only thing i don't like is how sensitive it would be to getting the order of `ee-impl` and `oss-impl` mixed up, and it's kinda hardcoded to just the two. If you're in there and want to change it a bit, maybe instead of `pred`, `ee-impl`, and `oss-impl` it could just take an impl-thunk which will return an impl. This allows for arbitrarily many impls in the future and makes sure that the macro invocation sites clearly express which impl goes with which branch of the pred.

```clojure
(ee-strategy-impl/reify-ee-strategy-impl #'settings.metastore/enable-sandboxes? impl params.field-values/default-impl
  params.field-values/FieldValuesForCurrentUser)

;; after
(ee-strategy-impl/reify-ee-strategy-impl #(if (settings.metastore/enable-sandboxes?) impl params.field-values/default-impl) params.field-values/FieldValuesForCurrentUser)
```

Reply:
that's a good idea

Reply:
Now that you brought this up, there's no reason this should really even be just an EE strategy impl anymore -- it could be called something more general like `strategy-impl` or `fn-strategy-impl` (not sure on the name). At that point moving the namespace and renaming everything and reworking all the macroexpansion tests would be a bit more that I want to tackle in this bugfix PR. I probably shouldn't have even made the pred fn change in the first place -- I thought I needed it at first but it turns out I didn't; the change still seemed like more correct behavior so I left it in place. I think the additional refactoring would make more sense in `master` as a follow-on PR